yash-executor/src/task.rs

```diff
@@ -0,0 +1,183 @@
+    /// this method panics.
+    pub fn poll(self: &Rc<Self>) -> bool {
+        if self.executor.strong_count() == 0 {
+            panic!("executor has been dropped");
+        }
+
+        let mut future_or_none = self
+            .future
+            .try_borrow_mut()
+            .expect("future polled recursively");
+        let Some(future) = future_or_none.as_mut() else {
+            return true;
+        };
+
+        let waker = into_waker(Rc::clone(self));
+        let mut context = Context::from_waker(&waker);
+        let poll = future.as_mut().poll(&mut context);
+        let is_ready = poll.is_ready();
+        if is_ready {
+            *future_or_none = None;
+        }
+        is_ready
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloc::boxed::Box;
+    use alloc::rc::Weak;
+    use core::cell::{Cell, RefCell};
+    use core::future::pending;
+
+    #[test]
+    fn waking_without_executor_does_nothing() {
+        let task = Rc::new(Task {
+            executor: Weak::new(),
+            future: RefCell::new(Some(Box::pin(async { unreachable!() }))),
+        });
+        task.wake();
+    }
+
+    #[test]
+    fn task_enqueues_itself_when_woken_with_executor() {
+        let executor = Rc::default();
+        let task = Rc::new(Task {
+            executor: Rc::downgrade(&executor),
+            future: RefCell::new(Some(Box::pin(async { unreachable!() }))),
+        });
+        assert_eq!(executor.borrow().wake_queue.len(), 0);
+
+        task.wake();
+        assert_eq!(executor.borrow().wake_queue.len(), 1);
+    }
+
+    #[test]
+    fn task_does_not_enqueue_again_if_already_enqueued() {
+        let executor = Rc::default();
+        let task = Rc::new(Task {
+            executor: Rc::downgrade(&executor),
+            future: RefCell::new(Some(Box::pin(async { unreachable!() }))),
+        });
+        Rc::clone(&task).wake();
+        assert_eq!(executor.borrow().wake_queue.len(), 1);
+
+        task.wake();
+        assert_eq!(executor.borrow().wake_queue.len(), 1);
+    }
+
+    #[test]
+    fn multiple_tasks_can_be_enqueued_at_once() {
+        let executor = Rc::default();
+        let task1 = Rc::new(Task {
+            executor: Rc::downgrade(&executor),
+            future: RefCell::new(Some(Box::pin(async { unreachable!() }))),
+        });
+        let task2 = Rc::new(Task {
+            executor: Rc::downgrade(&executor),
+            future: RefCell::new(Some(Box::pin(async { unreachable!() }))),
+        });
+
+        task1.wake();
+        task2.wake();
+        assert_eq!(executor.borrow().wake_queue.len(), 2);
+    }
+
+    #[test]
+    #[should_panic = "executor has been dropped"]
+    fn polling_without_executor_panics() {
+        let task = Rc::new(Task {
+            executor: Weak::new(),
+            future: RefCell::new(Some(Box::pin(async { unreachable!() }))),
+        });
+        task.poll();
+    }
```

Comment:
Polling without the executor cannot happen in normal use cases because it is the executor that calls the poll method. I don't think we need to handle such an anomaly. Returning an error value from the poll method would not work because the caller (the executor) would not be able to handle it anyway.